### ISSUE TITLE
Incorrect behavior of DateTimeToDateTimeInterfaceRector

### ISSUE THREAD

**Issue by puniserv** [rectorphp/rector#6644]
# Bug Report

| Subject        | Details              |
| :------------- | :--------------------|
| Rector version | last dev-master      |
| Installed as   | composer dependency  |

## Minimal PHP Code Causing Issue

See https://getrector.org/demo/1ebff5ab-027a-6aa4-8ff4-f5afbfaf26c6

```php
<?php

final class DemoFile
{
    private \DateTime $date;
    
    public function __construct(\DateTime $date){
		$this->date = $date;
	}

	public function getDate(): \DateTime
    {
		return $this->date;
	}
}
```

### Responsible rules

* `DateTimeToDateTimeInterfaceRector`

## Expected Behavior

Error:

```
TypeError : Typed property DemoFile::$date must be an instance of DateTime, DateTimeImmutable used
```

All types should be changed or nothing should be transformed.

**Comment by TomasVotruba** [comment]
Thank you for your report and demo link!

Could you send a failing test case in a pull-request, so we have it covered in Rector?
Here is step by step tutorial how to add it: https://github.com/rectorphp/rector/blob/master/docs/how_to_add_test_for_rector_rule.md

**Comment by Bl00D4NGEL** [comment]
I'm personally not sure if we want to apply this change to return types etc as well.

TL;DR: For typed properties it would be okay, for return types it'd probably cause more trouble than good.

Abstracting things usually results in having to check usage of the abstracted parameter. In case of DateTime|DateTimeImmutable -> DateTimeInterface a couple of functions will not be available anymore because we typehinted the interface (or any abstraction for the general discussion). An example would be the add function which exists on both implementations but not in the interface.

The rule in itself is quite risky imo since a lot of these "we abstract to the DateTimeInterface but actually need a specific implementation" cases can happen.

The "only" "good" way I can come up with is to add the @param/@return/@var doc to any replaced DateTime to DateTimeInterface, however I'm not sure if that is actually useful.

**Comment by TomasVotruba** [comment]
> The rule in itself is quite risky imo since a lot of these "we abstract to the DateTimeInterface but actually need a specific implementation" cases can happen.

In such case IMO the replace in whole class should be skipped.

**Comment by Bl00D4NGEL** [comment]
> In such case IMO the replace in whole class should be skipped.

So *if* we replace DateTime with DateTimeInterface everything should be replaced (just like @puniserv suggested), meaning that every typehint in the class should be changed from DateTime to DateTimeInterface, correct?

BUT we will only replace DateTime with DateTimeInterface if all usages of DateTime only refer to functions that are also used in the interface (like format, getOffet, getTimestamp) otherwise we will not replace anything, correct?

**Comment by TomasVotruba** [comment]
I'm not sure I understand.
Basically the replacement cannot break the code :)

**Comment by Wirone** [comment]
@puniserv In the end it was removed in [rectorphp/rector-src#6be87763](https://github.com/rectorphp/rector-src/commit/6be87763cbe27e0c20d3923df0efe629e5cae6f8) 😉 

<img width="848" alt="image" src="https://user-images.githubusercontent.com/600668/183474521-ea4244e1-aad9-444e-ac4e-0775e59c2bdf.png">

**Comment by TomasVotruba** [comment]
Yes, the behavior was rather experimental than based on practical solution of real problem. That's why it was so crappy :)
